dallinger/frontend/static/scripts/dallinger2.js

```diff
@@ -0,0 +1,295 @@
+      sParameterName,
+      i;
+
+    for (i = 0; i < sURLVariables.length; i++) {
+      sParameterName = sURLVariables[i].split('=');
+      if (sParameterName[0] === sParam) {
+        return sParameterName[1] === undefined ? true : sParameterName[1];
+      }
+    }
+  };
+
+  dlgr.identity = {
+    hitId: dlgr.getUrlParameter('hit_id'),
+    workerId: dlgr.getUrlParameter('worker_id'),
+    assignmentId: dlgr.getUrlParameter('assignment_id'),
+    mode: dlgr.getUrlParameter('mode'),
+    participantId: dlgr.getUrlParameter('participant_id')
+  };
+
+  dlgr.BusyForm = (function () {
+    /**
+    Loads a spinner as a visual cue that something is happening
+    and disables any jQuery objects passed to freeze().
+    **/
+
+    var defaults = {
+      spinnerSettings: {scale: 1.5}, // See http://spin.js.org/ for all settings
+      spinnerID: 'spinner'  // ID for HTML element where spinner will be inserted
+    };
+
+    var BusyForm = function (options) {
+      if (!(this instanceof BusyForm)) {
+        return new BusyForm(options);
+      }
+      var settings = $.extend(true, {}, defaults, options);
+      this.spinner = new Spinner(settings.spinnerSettings);
+      this.target = document.getElementById(settings.spinnerID);
+      if (this.target === null) {
+        throw new Error(
+          'Target HTML element for spinner with ID "' + settings.spinnerID +
+          '" does not exist.');
+      }
+      this.$elements = [];
+    };
+
+    BusyForm.prototype.freeze = function ($elements) {
+      this.$elements = $elements;
+      this.$elements.forEach(function ($element) {
+        $element.attr("disabled", true);
+      });
+      this.spinner.spin(this.target);
+    };
+
+    BusyForm.prototype.unfreeze = function () {
+      this.$elements.forEach(function ($element) {
+        $element.attr("disabled", false);
+      });
+      this.spinner.stop();
+      this.$elements = [];
+    };
+
+    return BusyForm;
+  }());
+
+  // stop people leaving the page, but only if desired by experiment
+  dlgr.allowExitOnce = false;
+  dlgr.preventExit = false;
+  window.addEventListener('beforeunload', function(e) {
+    if (dlgr.preventExit && !dlgr.allowExitOnce) {
+      var returnValue = "Warning: the study is not yet finished. " +
+        "Closing the window, refreshing the page or navigating elsewhere " +
+        "might prevent you from finishing the experiment.";
+      e.returnValue = returnValue;
+      return returnValue;
+    } else {
+      dlgr.allowExitOnce = false;
+      return undefined;
+    }
+  });
+  // allow actions to leave the page
+  dlgr.allowExit = function() {
+    dlgr.allowExitOnce = true;
+  };
+
+  // advance the participant to a given html page
+  dlgr.goToPage = function(page) {
+    window.location = "/" + page + "?participant_id=" + dlgr.identity.participantId;
+  };
+
+  // AJAX helpers
+
+  var ajax = function (method, route, data) {
+    var deferred = $.Deferred();
```

Comment:
I guess it's really just a more standardized way of using callback functions, that I find convenient. Another option would be to use ES6 promises, but jQuery Deferred is similar and was already available. I think Jesse used it because I used it for the waiting room.